libp2p/protocols/secure/noise.nim

```diff
@@ -0,0 +1,575 @@
+  await sconn.readExactly(addr besize[0], 2)
+  bigEndian16(addr size, addr besize[0])
+  trace "receiveEncryptedMessage", size, peer = $sconn.peerInfo
+  if size == 0:
+    return @[]
+  let
+    cipher = await sconn.read(size.int)
+  var plain = sconn.readCs.decryptWithAd([], cipher)
+  unpackNoisePayload(plain)
+  return plain
+
+proc sendEncryptedMessage(sconn: NoiseConnection; buf: seq[byte]) {.async.} =
+  let
+    packed = packNoisePayload(buf)
+    cipher = sconn.writeCs.encryptWithAd([], packed)
+  var
+    lesize = cipher.len
+    size: array[2, byte]
+  trace "sendEncryptedMessage", size = lesize, peer = $sconn.peerInfo
+  bigEndian16(addr size[0], addr lesize)
+  await sconn.write(addr size[0], 2)
+  await sconn.write(cipher)
+  
+proc handshakeXXOutbound(p: Noise, conn: Connection, p2pProof: ProtoBuffer): Future[HandshakeResult] {.async.} =
+  const initiator = true
+
+  var
+    hs: HandshakeState
+    empty: seq[byte]
+  init hs.ss
+
+  var p2psecret = p2pProof.buffer
+
+  hs.ss.mixHash(p.commonPrologue)
+  hs.s.privateKey = p.noisePrivateKey
+  hs.s.publicKey = p.noisePublicKey
+
+  # -> e
+  var msg: seq[byte]
+
+  write_e()
+
+  # IK might use this btw!
+  msg &= hs.ss.encryptAndHash(empty)
+
+  await conn.sendHSMessage(msg)
+
+  # <- e, ee, s, es
+
+  msg = await conn.receiveHSMessage()
+
+  read_e()
+  dh_ee()
+  read_s()
+  dh_es()
+
+
+  var remoteP2psecret = hs.ss.decryptAndHash(msg)
+  unpackNoisePayload(remoteP2psecret)
+
+  # -> s, se
+
+  msg.setLen(0)
+
+  write_s()
+  dh_se()
+
+  # last payload must follow the ecrypted way of sending
+  var packed = packNoisePayload(p2psecret)
+  msg &= hs.ss.encryptAndHash(packed)
+
+  await conn.sendHSMessage(msg)
+  
+  let (cs1, cs2) = hs.ss.split()
+  return HandshakeResult(cs1: cs1, cs2: cs2, remoteP2psecret: remoteP2psecret, rs: hs.rs)
+
+proc handshakeXXInbound(p: Noise, conn: Connection, p2pProof: ProtoBuffer): Future[HandshakeResult] {.async.} =
+  const initiator = false
+
+  var
+    hs: HandshakeState
+    empty: seq[byte]
+  init hs.ss
+
+  var p2psecret = p2pProof.buffer
+
+  hs.ss.mixHash(p.commonPrologue)
+  hs.s.privateKey = p.noisePrivateKey
+  hs.s.publicKey = p.noisePublicKey
+
+  # -> e
+
+  var msg = await conn.receiveHSMessage()
+
+  read_e()
+
+  let earlyData = hs.ss.decryptAndHash(msg)
+
+  # <- e, ee, s, es
+
+  msg.setLen(0)
+
+  write_e()
+  dh_ee()
+  write_s()
+  dh_es()
+
+  var packedSecret = packNoisePayload(p2psecret)
+  msg &= hs.ss.encryptAndHash(packedSecret)
+
+  await conn.sendHSMessage(msg)
+
+  # -> s, se
+
+  msg = await conn.receiveHSMessage()
+
+  read_s()
+  dh_se()
+
+  var remoteP2psecret = hs.ss.decryptAndHash(msg)
+  unpackNoisePayload(remoteP2psecret)
+
+  let (cs1, cs2) = hs.ss.split()
+  return HandshakeResult(cs1: cs1, cs2: cs2, remoteP2psecret: remoteP2psecret, rs: hs.rs)
+
+proc readMessage(sconn: NoiseConnection): Future[seq[byte]] =
+  try:
+    return sconn.receiveEncryptedMessage()
+  except AsyncStreamIncompleteError:
```

Comment:
why does this drop the exception?

Reply:
Not sure, copied this bit of code from `secio` as it's something shared (we might actually move it into secure and really share)
All of the codebase is quite confused  in terms of exceptions and error handling and needs a major review.

Reply:
so `secio` most likely is buggy the same way and needs fixing

Reply:
ping @dryajov what do you think?

Reply:
Will do that after merging cos Secio needs this too